### PR TITLE
Add support for HTTP cookies

### PR DESCRIPTION
**Pull request by julienr**
This is based on initial work by @awaterme in #345, which was closed due to inactivity. I'm happy to take over and try to get this merged as we have the same need as Andreas.

This is the same code as #345, but I've addressed the comments about documentation and squashed all the commits. If you prefer that I leave Andreas as the author of the squashed commit, I can also do that (my intent is not to steal any credits, so please tell me what's best). Or @awaterme, if you actually want to continue working on your PR, happy to help you then.

This enables for sources that require cookies to manage session (for example rate-limiting clients based on number of sessions opened).

### CODE REVIEW

File: doc/configuration.rst [hunk 1]
```diff
@@ -1024,6 +1024,15 @@ Add additional HTTP headers to all requests to your sources.
 
 Sets the ``Access-control-allow-origin`` header to HTTP responses for `Cross-origin resource sharing <http://en.wikipedia.org/wiki/Cross-origin_resource_sharing>`_. This header is required for WebGL or Canvas web clients. Defaults to `*`. Leave empty to disable the header. This option is only available in `globals`.
 
+``manage_cookies``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. versionadded:: 1.12.0
```
Comment (weskamm) [inline doc/configuration.rst:1030]: should read 1.14.0 (as the current latest version is 1.13)

**Comment by awaterme** [comment]
Hi Julien,
I'm very glad somebody picked up the issue again. Please go ahead  and don't worry about my credits.
Best regards,
Andreas

File: mapproxy/config_template/base_config/full_example.yaml [hunk 2]
```diff
@@ -425,10 +425,27 @@ sources:
     req:
       # username and password are extracted from the URL and do not show
       # up in log files
+      # ATTENTION: see https://github.com/mapproxy/mapproxy/issues/182
```
Comment (weskamm) [inline mapproxy/config_template/base_config/full_example.yaml:428]: please remove this line and move the information to the docs in the appropriate section

File: mapproxy/config_template/base_config/full_example.yaml [hunk 3]
```diff
@@ -425,10 +425,27 @@ sources:
...
 
+  # WMS source that requires authentication and session management
+  # through HTTP cookies
+  session_source:
+    type: wms
+    http:
+      # Accept session cookies and forward on subsequent requests
+      manage_cookies: true
+      # Use basic auth header directly
+      # see https://github.com/mapproxy/mapproxy/issues/182
```
Comment (weskamm) [inline mapproxy/config_template/base_config/full_example.yaml:441]: same as above, please remove this line

**Comment by weskamm** [comment]
Thanks for picking this up again.
Just found 2 minor issues, which you may want to adress. Else this is looking good!

File: doc/configuration.rst [hunk 4]
```diff
@@ -1024,6 +1024,15 @@ Add additional HTTP headers to all requests to your sources.
 
 Sets the ``Access-control-allow-origin`` header to HTTP responses for `Cross-origin resource sharing <http://en.wikipedia.org/wiki/Cross-origin_resource_sharing>`_. This header is required for WebGL or Canvas web clients. Defaults to `*`. Leave empty to disable the header. This option is only available in `globals`.
 
+``manage_cookies``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (olt) [inline doc/configuration.rst:1028]: .rst requires that the ^^^ line length matches the title.

File: doc/configuration_examples.rst [hunk 5]
```diff
@@ -734,11 +736,28 @@ You can also set this in your :ref:`server script <server_script>`::
...
 
+.. note:: If the source requires session handling through cookies, have a look at the ``manage_cookies`` configuration option.
+
 You can use the ``no_proxy`` environment variable if you need to bypass the proxy for some hosts::
 
   $ export no_proxy="localhost,127.0.0.1,196.168.1.99"
 
 
+Cookie Management
+==================
```
Comment (olt) [inline doc/configuration_examples.rst:747]: Same here.

**Review by olt — COMMENTED** [review]
A feature like this should have at least a small system test to assure that it does work now and more importantly in the future.

**Comment by julienr** [comment]
Thanks for the reviews. I've added a test for the http client behavior of Set-Cookie. I'll just add another one in `test_conf_loader` checking that the config parameter is correctly passed from config file to http client.

File: doc/configuration.rst (hunk 1 above)
Comment (julienr) [reply doc/configuration.rst]: Good catch - I've fixed it

File: mapproxy/config_template/base_config/full_example.yaml (hunk 2 above)
Comment (julienr) [reply mapproxy/config_template/base_config/full_example.yaml]: Removed

File: mapproxy/config_template/base_config/full_example.yaml (hunk 3 above)
Comment (julienr) [reply mapproxy/config_template/base_config/full_example.yaml]: Removed

**Comment by julienr** [comment]
> A feature like this should have at least a small system test to assure that it does work now and more importantly in the future.

I've added a system test to `test_wms` and two unit test to `test_client.py` (to check that enabling/disabling the functionality has the correct behavior). 

Is it the kind of tests you were expecting or should I do it differently ?

File: doc/configuration.rst (hunk 4 above)
Comment (julienr) [reply doc/configuration.rst]: Good catch - fixed

File: doc/configuration_examples.rst (hunk 5 above)
Comment (julienr) [reply doc/configuration_examples.rst]: Good catch - fixed

**Review by weskamm — COMMENTED** [review]
The tests are failing, but it looks like its easy to fix.
I think the idea of testing on and off behaviour is good.

I am a little bit confused about the structure of your tests, as it looks somewhat different to all the other tests. But this is maybe just a personal taste.
So if the tests pass again, i think this could be merged. Thanks!

**Comment by julienr** [comment]
@weskamm I've fixed the tests - sorry for pushing something broken.

As far as test structure go, are you referring to the `assert_cookie` / `assert_no_cookie` functions ? I've done it that way (passing a helper to the mock request setup) because that's the only way I found to access the request from the mock server side. But if there is a better way, I'd be happy to do it.

**Comment by weskamm** [comment]
I was hoping that someone else could answer to that question, but i think its ok. Thanks!

**Review by weskamm — APPROVED** [review]

**Pull request merged**